File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/FunctionName.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Lexy.Poc.Core.Parser;
using Microsoft.CodeAnalysis.CSharp;

namespace Lexy.Poc.Core.Language
{
    public class FunctionName : Node
    {
        public string Value { get; private set; }

        public FunctionName(SourceReference reference) : base(reference)
        {
        }

        public void ParseName(string parameter = null)
        {
            Value = parameter ?? "Function__" + Guid.NewGuid().ToString("N");
        }

        public string ClassName()
        {
            var nameBuilder = new StringBuilder("Function");
            foreach (var @char in Value.Where(char.IsLetter))
            {
                nameBuilder.Append(@char);
            }
            return nameBuilder.ToString();
        }

        protected override IEnumerable<INode> GetChildren()
        {
            yield break;
        }

        protected override void Validate(IValidationContext context)
        {
            if (string.IsNullOrEmpty(Value))
            {
                context.Logger.Fail(Reference, $"Invalid function name: {Value}. Name should not be empty.");
            }
            if (!SyntaxFacts.IsValidIdentifier(Value))
            {
                context.Logger.Fail(Reference, $"Invalid function name: {Value}.");
            }
        }
    }
}

File: typescipt/src/language/functions/functionName.ts


export class FunctionName extends Node {
   public string Value { get; private set; }

   public FunctionName(SourceReference reference) : base(reference) {
   }

   public parseName(name: string): void {
     Value = name ?? throw new Error(nameof(name));
   }

   public override getChildren(): Array<INode> {
     yield break;
   }

   protected override validate(context: IValidationContext): void {
     if (string.IsNullOrEmpty(Value))
       context.Logger.Fail(Reference, $`Invalid function name: '{Value}'. Name should not be empty.`);
     if (!SyntaxFacts.IsValidIdentifier(Value)) context.Logger.Fail(Reference, $`Invalid function name: '{Value}'.`);
   }
}
